pyexchange/kucoin.py

```diff
@@ -0,0 +1,269 @@
+                     pair=pair,
+                     is_sell=trade['direction'] == 'SELL',
+                     price=Wad.from_number(trade['dealPrice']),
+                     amount=Wad.from_number(trade['amount']))
+
+    @staticmethod
+    def from_list(pair, trade):
+        # [1544564526000, 'SELL', 25.0005, 0.0614088, 1.5352507, '5c102f2d335e7e08134edd77']
+        return Trade(trade_id=None,
+                     order_id=trade[5],
+                     timestamp=float(trade[0]),
+                     pair=pair,
+                     is_sell=trade[1] == 'SELL',
+                     price=Wad.from_number(trade[2]),
+                     amount=Wad.from_number(trade[3]))
+
+class KucoinApi(PyexAPI):
+    """kucoin API interface.
+    """
+
+    logger = logging.getLogger()
+
+    def __init__(self, api_server: str, api_key: str, secret_key: str, timeout: float, requests_params=None):
+        assert(isinstance(api_server, str))
+        assert(isinstance(api_key, str))
+        assert(isinstance(secret_key, str))
+        assert(isinstance(timeout, float))
+
+        self.api_server = api_server
+        self.api_key = api_key
+        self.secret_key = secret_key
+        self.timeout = timeout
+        self.client = Client(api_key, secret_key)
+
+    def get_markets(self):
+        return self.client.get_trading_markets()
+
+    def ticker(self, pair: str):
+        assert(isinstance(pair, str))
+        return self.client.get_tick(pair);
+
+    def get_balances(self):
+        return self.client.get_all_balances()
+
+    def get_fiat_balance(self, fiat : str):
+        assert(isinstance(fiat, str))
+        return self.client.get_total_balance(fiat)
+
+    def get_balance(self, coin : str):
+        assert(isinstance(coin, str))
+        return self.client.get_coin_balance(coin)
+
+    def get_user_info(self):
+        return self.client.get_user()
+
+    def order_book(self, pair: str, limit=None):
+        assert(isinstance(pair, str))
+        return self.client.get_order_book(pair, limit)
+
+    def get_orders(self, pair: str)  -> List[Order]:
+        assert(isinstance(pair, str))
+
+        orders = self.client.get_active_orders(pair)
+
+        sell_orders = list(map(lambda item: Order.from_list(item, pair, True), orders[self.client.SIDE_SELL]))
+        buy_orders = list(map(lambda item: Order.from_list(item, pair, False), orders[self.client.SIDE_BUY]))
+
+        return sell_orders + buy_orders
+
+    def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> str:
+        assert(isinstance(pair, str))
+        assert(isinstance(is_sell, bool))
+        assert(isinstance(price, Wad))
+        assert(isinstance(amount, Wad))
+
+        side = self.client.SIDE_SELL if is_sell else self.client.SIDE_BUY
+
+        coins = pair.split("-")
+
+        price = self._get_precision(coins[1]) % float(price)
+        amount = self._get_precision(coins[0]) % float(price)
```

Comment:
are you sure it shouldn't say `float(amount)` here??

Reply:
API rejects if they do not follow an exact precision, I determined precision by testing different pairs as didn't find any documentation for